crates/editor/src/editor.rs

```diff
@@ -4722,6 +4726,93 @@ impl Editor {
         None
     }
 
+    pub fn refresh_selected_text_highlights(
+        &mut self,
+        window: &mut Window,
+        cx: &mut Context<Editor>,
+    ) {
+        self.selection_highlight_task.take();
+        if !EditorSettings::get_global(cx).selection_highlight {
+            self.clear_background_highlights::<SelectedTextHighlight>(cx);
+            return;
+        }
+        if self.selections.count() != 1 || self.selections.line_mode {
+            self.clear_background_highlights::<SelectedTextHighlight>(cx);
+            return;
+        }
+        let selection = self.selections.newest::<Point>(cx);
+        if selection.is_empty() || selection.start.row != selection.end.row {
+            self.clear_background_highlights::<SelectedTextHighlight>(cx);
+            return;
+        }
+        let debounce = EditorSettings::get_global(cx).selection_highlight_debounce;
+        self.selection_highlight_task = Some(cx.spawn_in(window, |editor, mut cx| async move {
+            cx.background_executor()
+                .timer(Duration::from_millis(debounce))
+                .await;
+            let Some(matches_task) = editor
+                .read_with(&mut cx, |editor, cx| {
+                    let buffer = editor.buffer().read(cx).snapshot(cx);
+                    cx.background_executor().spawn(async move {
+                        let mut ranges = Vec::new();
+                        let buffer_ranges =
+                            vec![buffer.anchor_before(0)..buffer.anchor_after(buffer.len())];
+                        let query = buffer.text_for_range(selection.range()).collect::<String>();
+                        for range in buffer_ranges {
```

Comment:
This is a bit odd: we create a non-mut `vec![buffer.anchor_before(0)..buffer.anchor_after(buffer.len())]` and then iterate it?
Can unwrap the entire loop it seems.

------------------

One potential super nice feature could be expanding a single, empty selection (basically, a caret) to the ends of the word with `movement::surrounding_word`: then, we can implement highlights like

<img width="1155" alt="image" src="https://github.com/user-attachments/assets/b25c4403-3c8f-4101-953d-a88991bb18c0" />

which we do with tree-sitter.